Add User Endpoints

Matthew-Grayson opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release.

Matthew-Grayson committed:
Add delete_user endpoint.

Matthew-Grayson committed:
Add Accept Terms Endpoint.

Add accept_terms function to api_methods/user.py
Add /users/acceptTerms endpoint to views.py
Add UpdateUser class to schema_models/user.py
Add can_access_user function to auth.py

Matthew-Grayson committed:
Modify and reorder delete user endpoint; add todos for authentication and permissions.

Matthew-Grayson committed:
Add Update Users Endpoint

Add update users endpoint to views
Add update users definition to api_methods/user.py
Refactor user endpoints to use Request object for passing arguments

Matthew-Grayson committed:
Merge branch 'AL-python-serverless' of github.com:cisagov/XFD into MG-python-serverless

Matthew-Grayson committed:
Update User Models and Endpoints; Update Input Validation for Update User Endpoints.

Matthew-Grayson committed:
Add send_email Helper Functions.

Matthew-Grayson committed:
Merge branch 'AL-python-serverless' of github.com:cisagov/XFD into MG-python-serverless

Matthew-Grayson committed:
Add get users by state and get users by region endpoints.

aloftus23 committed:
Add register endpoints and post User with tests in test_user.py

Matthew-Grayson committed:
Update python user endpoints.

Matthew-Grayson committed:
Merge origin MG-python-serverless into remote MG-python-serverless.

Matthew-Grayson committed:
Merge origin MG-python-serverless into remote MG-python-serverless.

Matthew-Grayson committed:
Fix lint issues in email.py.

Matthew-Grayson committed:
Merge branch 'AL-python-serverless' of github.com:cisagov/XFD into MG-python-serverless

Matthew-Grayson committed:
Merge branch 'AL-python-serverless' of github.com:cisagov/XFD into MG-python-serverless

cduhn17 commented on backend/src/xfd_django/xfd_api/api_methods/user.py:
```diff
@@ -3,30 +3,482 @@
+        raise HTTPException(status_code=500, detail=str(e))
+
+
+#  TODO: Add user context and permissions
```
Create ticket associated with all TODO messages for tracking purposes

cduhn17 commented on backend/src/xfd_django/xfd_api/api_methods/user.py:
```diff
@@ -3,30 +3,482 @@
+    """Deny a user's registration by user ID."""
+    try:
+        # Validate UUID format for the user_id
+        if not is_valid_uuid(user_id):
```
For consistency for logic checks on `is_valid_uuid` should be outside of try blocks like shown on `def user_register_approve`. In general anything that does not involve i.e. database connections or services connections like email(this is not an exhaustive list) should be outside of try blocks. Things that qualify for in try block are exception prone operations.

cduhn17 commented on backend/src/xfd_django/xfd_api/helpers/email.py:
```diff
@@ -0,0 +1,153 @@
+def send_invite_email(email, organization=None):
+    """Send an invitation email to the specified address."""
+    frontend_domain = os.getenv("FRONTEND_DOMAIN")
+    reply_to = os.getenv("CROSSFEED_SUPPORT_EMAIL_REPLYTO")
```
if you want to drive constants from the settings file
`import django.conf import settings`. Then you can change os.getenv to `settings.("CROSSFEED_SUPPORT_EMAIL_REPLYTO")`. The reasoning is then constants wont be overridden by test mocks and to not depend on the environment to get constants. Django caches constants for efficient retrieval as well on start up.

cduhn17 reviewed (commented):
See comments.

Matthew-Grayson committed:
Refactor User Endpoints to Follow Best Practice

Add user context and permissions to delete_user and remove associated todo
Move non-data/service connections outside of try blocks
Replace os.getenv() calls with settings() in email.py
Add regional admin permission checks to can_access_user in auth.py

Matthew-Grayson committed:
Merge branch 'AL-python-serverless' of github.com:cisagov/XFD into MG-python-serverless

Matthew-Grayson committed:
Revert Deleted User Endpoints

nickviola reviewed (approved):
I'm not sure if we still need the invitation email stuff, but we can address later in a SPIKE ticket. Otherwise this LGTM

aloftus23 reviewed (approved):
Looks good to me
